File: Pieces/king.py
from Pieces.piece import Piece
from Pieces.moveCandidate import MoveCandidate
from enums import *
from coords import Coords

class King(Piece):
    def __init__(self, colour: Colour) -> None:
        super().__init__(colour)

    def get_representation(self) -> str:
        return 'k' if self.colour == Colour.BLACK else 'K'

    def get_candidate_moves(self, coords: Coords):
        # list of tuples of new coordinates the piece can go
        candidates = []

        for di, dj in [
            [-1, -1], [-1, 1], [1, -1], [1, 1],
            [-1, 0], [1, 0], [0, -1], [0, 1]
            ]:

            dist = 1

            # castling
            if (str(coords) == "e1" or str(coords) == "e8") and di == 0:
                dist = 2
            candidate = MoveCandidate(di, dj, dist)
            candidates.append(candidate)


        return candidates

File: rust_chess/src/pieces/king.rs
use std::fmt::DebugSet;

use crate::enums::Colour;
use crate::coords::Coords;
use crate::moves::move_ray::MoveRay;
use crate::pieces::Piece;

#[derive(Debug, Clone)]
pub struct King {
    pub colour: Colour,
    pub position: Coords,
}

impl Piece for King {
    fn colour(&self) -> Colour {
        self.colour
    }
    fn get_representation(&self) -> char {
        match self.colour {
            Colour::White => 'K',
            Colour::Black => 'k',
        }
    }
    fn get_move_rays(&self, from: Coords) -> Vec<MoveRay> {
        let mut destinations = Vec::new();
        for (rank_diff, file_diff) in &[
            (-1, -1), (-1, 0), (-1, 1),
            (0, -1),          (0, 1),
            (1, -1),  (1, 0),  (1, 1),
        ] {
            if let Some(new_coords) = from.get_neighbour(*rank_diff, *file_diff) {
                let move_ray = MoveRay::new(*rank_diff, *file_diff, 1, true, false);
                destinations.push(move_ray);
            }
        }

        return destinations;
    }
}